ledger/accountdb.go

```diff
@@ -192,24 +195,73 @@ const (
+	normalizedAccountBalances = make([]normalizedAccountBalance, len(bals))
+	for i, balance := range bals {
+		normalizedAccountBalances[i].address = balance.Address
+		var accountDataV5 basics.AccountData
+		err = protocol.Decode(balance.AccountData, &accountDataV5)
+		if err != nil {
+			return nil, err
+		}
+		normalizedAccountBalances[i].accountData.SetAccountData(&accountDataV5)
+		normalizedAccountBalances[i].normalizedBalance = accountDataV5.NormalizedOnlineBalance(proto)
+		resources := accountDataResources(&accountDataV5)
+		normalizedAccountBalances[i].accountHashes = make([][]byte, 1)
+		normalizedAccountBalances[i].accountHashes[0] = accountHashBuilder(balance.Address, accountDataV5, balance.AccountData)
+		if len(resources) > 0 {
+			normalizedAccountBalances[i].resources = make(map[basics.CreatableIndex]resourcesData, len(resources))
+		}
+		for _, resource := range resources {
+			normalizedAccountBalances[i].resources[resource.aidx] = resource.resourcesData
+		}
+		normalizedAccountBalances[i].encodedAccountData = balance.AccountData
+
+	}
+	return
+}
+
+// prepareNormalizedBalancesV6 converts an array of encodedBalanceRecordV5 into an equal size array of normalizedAccountBalances.
+func prepareNormalizedBalancesV6(bals []encodedBalanceRecordV6, proto config.ConsensusParams) (normalizedAccountBalances []normalizedAccountBalance, err error) {
+	normalizedAccountBalances = make([]normalizedAccountBalance, len(bals))
 	for i, balance := range bals {
 		normalizedAccountBalances[i].address = balance.Address
 		err = protocol.Decode(balance.AccountData, &(normalizedAccountBalances[i].accountData))
 		if err != nil {
 			return nil, err
 		}
-		normalizedAccountBalances[i].normalizedBalance = normalizedAccountBalances[i].accountData.NormalizedOnlineBalance(proto)
+		normalizedAccountBalances[i].normalizedBalance = basics.NormalizedOnlineAccountBalance(
+			normalizedAccountBalances[i].accountData.Status,
+			normalizedAccountBalances[i].accountData.RewardsBase,
+			normalizedAccountBalances[i].accountData.MicroAlgos,
+			proto)
 		normalizedAccountBalances[i].encodedAccountData = balance.AccountData
-		normalizedAccountBalances[i].accountHash = accountHashBuilder(balance.Address, normalizedAccountBalances[i].accountData, balance.AccountData)
+		normalizedAccountBalances[i].accountHashes = make([][]byte, 1+len(balance.Resources))
+		normalizedAccountBalances[i].accountHashes[0] = accountHashBuilderV6(balance.Address, &normalizedAccountBalances[i].accountData, balance.AccountData)
+		normalizedAccountBalances[i].resources = make(map[basics.CreatableIndex]resourcesData, len(balance.Resources))
+		resIdx := 0
+		for cidx, res := range balance.Resources {
+			var resData resourcesData
+			err = protocol.Decode(res, &resData)
+			if err != nil {
+				return nil, err
+			}
+			ctype := basics.AssetCreatable
+			if resData.IsEmptyApp() {
+				ctype = basics.AppCreatable
+			}
```

Comment:
why does `resData.IsEmptyApp()` trigger ctype = basics.AppCreatable ? I guess the resolution must be done on `ResourceFlags`.

Reply:
exactly. We don't store the asset/app differentiation anywhere else, which is why we need to use the ResourceFlags to figure this out.